### ISSUE TITLE
Removing line marks from the scenegraph once thier datum is removed.

### ISSUE THREAD

**Issue by rmcgregor1990** [vega/vega#1305]
Hey,

I'm trying to make a live line plot with a scrolling window of the last 30 seconds on the x axis.
My spec looks like this:
```json
{
  "$schema": "https://vega.github.io/schema/vega/v3.json",
  "autosize": "pad",
  "padding": 20,
  "signals": [
    {
      "name": "xDomain",
      "value": [
        0,
        0
      ]
    }
  ],
  "data": [
    {
      "name": "livedata"
    }
  ],
  "scales": [
    {
      "name": "time",
      "type": "time",
      "range": "width",
      "domain": {
        "signal": "xDomain"
      }
    },
    {
      "name": "y",
      "type": "linear",
      "range": "height",
      "domain": {
        "data": "livedata",
        "field": "y"
      }
    }
  ],
  "axes": [
    {
      "orient": "bottom",
      "scale": "time"
    },
    {
      "orient": "left",
      "scale": "y"
    }
  ],
  "marks": [
    {
      "type": "line",
      "from": {
        "data": "livedata"
      },
      "clip": true,
      "encode": {
        "update": {
          "x": {
            "scale": "time",
            "field": "time"
          },
          "y": {
            "scale": "y",
            "field": "y"
          },
          "defined": {
            "signal": "datum.time > xDomain[0]"
          }
        }
      }
    }
  ]
}
```
I'm achieving the live time scrolling animation by updating the ```xDomain``` signal with
```javascript
view.signal("xDomain", [Date.now() - 30e3 -200, Date.now() - 200]);
``` 
via the view API and calling .run() in the browser ```requestAnimationFrame``` callback.

The backing data source ```livedata``` is updated every ~100ms via the view API call:
```javascript
view.change('livedata', view.changeset().remove(()=>true).insert(livedata))
```
Here the ```livedata``` variable only contains the last ~30sec of the stream with old data being dropped.

This initially works pretty well, the data is smoothly plotted over time (which each re-render taking ~4ms on my machine).
However when left for any period of time it becomes very sluggish with the render time steadily increasing. I believe this is due to the marks on the scene-graph not being removed even after the data backing them has gone. Hence they build up over time and bring the renderer to its knees.

So my Question is:
Is there any way to removed marks from the scene-graph after their datum has gone?

I'm basically looking for something akin to d3s
```javascript
.exit().remove()
```
I however not find anything in the documentation regarding this.

Cheers,
Rob

**Comment by kenklin** [comment]
I'm interested in this thread as I will be wiring up my data to be live (1Hz).

**Comment by rmcgregor1990** [comment]
Yeah I'm pretty sure its the marks not being pruned.
Here's a graph of length of the data source vs the length of the marks array within the scenegraph:
![marks](https://user-images.githubusercontent.com/7824682/40626652-e097c7bc-62b1-11e8-8d4a-d94352d5e88c.png)

**Comment by jheer** [comment]
Thanks for the issue! Vega is intended to support streaming data.

Internally, the following things should be happening:
- When you remove data, the corresponding mark items should be removed from the scenegraph.
- As items are often filtered in/out due to interaction, we also cache items internally to quickly add them back to the scenegraph as needed.
- Once a cache reaches a threshold size (determined by the `view.cleanThreshold` property, which defaults to 10000), currently unused entries are cleared, reclaiming memory.

To test, I created my own variant of your code above -- expand the collapsed block below to see the full HTML file. Using the latest Vega version (4.0.0-rc.2), I'm unable to replicate the problem.

The test code prints diagnostics to the console with every 100ms data update interval, tracking the number of data tuples, line mark items in the scenegraph, line mark entries in the DataJoin cache, and the true elapsed time between updates. After running for over 15 minutes, all diagnostics are still behaving as expected: the data and item counts are constant, the cache size increases up to 10000 then resets due to garbage collection, and the time between updates stays near 100ms.

Please let me know if you see the same behavior once updating to the latest Vega version (if needed). Also let me know if your actual code or spec differs in a meaningful way from the test code here. If you are still seeing performance issues, please share additional details, including version, OS, browser, etc info. If you share a link to an online example that exhibits the problem that would be helpful as well.

<details>
<summary>Click to expand: Streaming Test HTML Page</summary><p>

```html
<!DOCTYPE HTML>
<html>
  <head>
    <title>Vega Streaming Data Test</title>
    <script src="https://d3js.org/d3.v5.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/vega@4.0.0-rc.2"></script>
  </head>
  <body>
    <div id="view"></div>
    <script>
var runtimeSpec = vega.parse({
  "$schema": "https://vega.github.io/schema/vega/v4.json",
  "autosize": "pad",
  "padding": 20,
  "width": 500,
  "height": 200,
  "signals": [
    {
      "name": "xDomain",
      "value": [
        0,
        0
      ]
    }
  ],
  "data": [
    {
      "name": "livedata"
    }
  ],
  "scales": [
    {
      "name": "time",
      "type": "time",
      "range": "width",
      "domain": {
        "signal": "xDomain"
      }
    },
    {
      "name": "y",
      "type": "linear",
      "range": "height",
      "domain": {
        "data": "livedata",
        "field": "y"
      }
    }
  ],
  "axes": [
    {
      "orient": "bottom",
      "scale": "time"
    },
    {
      "orient": "left",
      "scale": "y"
    }
  ],
  "marks": [
    {
      "type": "line",
      "from": {
        "data": "livedata"
      },
      "clip": true,
      "encode": {
        "update": {
          "x": {
            "scale": "time",
            "field": "time"
          },
          "y": {
            "scale": "y",
            "field": "y"
          },
          "defined": {
            "signal": "datum.time > xDomain[0]"
          }
        }
      }
    }
  ]
});

var view = new vega.View(runtimeSpec)
  .initialize(document.querySelector('#view'))
  .renderer('svg')
  .hover()
  .run();

// Adjust this value to determine size threshold at
// which to perform garbage cleaning of internal caches
// The default value is 10000
// view.cleanThreshold = 500;

var count = 0,
    t0 = Date.now();

// update axis domain every render frame
function update() {
  view.signal("xDomain", [Date.now() - 30e3 -200, Date.now() - 200]);
  view.run();
  window.requestAnimationFrame(update);
}

// update data every 100ms
setInterval(function() {
  var domain = view.signal('xDomain'),
      step = (domain[1] - domain[0]) / 20;

  var livedata = d3
    .range(domain[0], domain[1] + step, step)
    .map(function(d) { return {time: d, y: Math.random()}; });

  view
    .change('livedata', view.changeset()
      .remove(() => true)
      .insert(livedata))
    .run();

  console.log(
    'ITER',  ++count,
    'DATA',  view._runtime.data.livedata.values.value.length,
    'ITEMS', view.scenegraph().root.items[0].items[2].items.length,
    'CACHE', view._runtime.nodes[79].value.size,
    'TIME',  Date.now() - t0
  );
  t0 = Date.now();
}, 100);

update();
    </script>
  </body>
</html>
```

</p></details>

**Comment by rmcgregor1990** [comment]
Thanks for looking into it!

It seems to be related to the actual data used to update the view each tick. 
When new data arrives I'm slicing the existing dataset remove old datums outside the x axis and then appending the fresh data. I then overwrite the views data with this updated dataset via the changeset.
However the marks from these old datums are not being removed from the scenegraph.

I've modified your example to this effect and you can see the scenegraph items not being culled once the 30 sec window is filled.
Sorry I don't know how to do your click to expand markdown magic.

```html
<!DOCTYPE HTML>
<html>
  <head>
    <title>Vega Streaming Data Test</title>
    <script src="https://d3js.org/d3.v5.min.js"></script>
    <script src="https://cdn.jsdelivr.net/npm/vega@4.0.0-rc.2"></script>
  </head>
  <body>
    <div id="view"></div>
    <script>
var runtimeSpec = vega.parse({
  "$schema": "https://vega.github.io/schema/vega/v4.json",
  "autosize": "pad",
  "padding": 20,
  "width": 500,
  "height": 200,
  "signals": [
    {
      "name": "xDomain",
      "value": [
        0,
        0
      ]
    }
  ],
  "data": [
    {
      "name": "livedata"
    }
  ],
  "scales": [
    {
      "name": "time",
      "type": "time",
      "range": "width",
      "domain": {
        "signal": "xDomain"
      }
    },
    {
      "name": "y",
      "type": "linear",
      "range": "height",
      "domain": {
        "data": "livedata",
        "field": "y"
      }
    }
  ],
  "axes": [
    {
      "orient": "bottom",
      "scale": "time"
    },
    {
      "orient": "left",
      "scale": "y"
    }
  ],
  "marks": [
    {
      "type": "line",
      "from": {
        "data": "livedata"
      },
      "clip": true,
      "encode": {
        "update": {
          "x": {
            "scale": "time",
            "field": "time"
          },
          "y": {
            "scale": "y",
            "field": "y"
          },
          "defined": {
            "signal": "datum.time > xDomain[0]"
          }
        }
      }
    }
  ]
});

var view = new vega.View(runtimeSpec)
  .initialize(document.querySelector('#view'))
  .renderer('svg')
  .hover()
  .run();

// Adjust this value to determine size threshold at
// which to perform garbage cleaning of internal caches
// The default value is 10000
// view.cleanThreshold = 500;

var count = 0,
    t0 = Date.now();

// update axis domain every render frame
function update() {
  view.signal("xDomain", [Date.now() - 30e3 -200, Date.now() - 200]);
  view.run();
  window.requestAnimationFrame(update);
}

var data = [{y:0, time: Date.now()}];

// update data every 100ms
setInterval(function() {
  var domain = view.signal('xDomain'),
      step = (domain[1] - domain[0]) / 20;

  view
    .change('livedata', view.changeset()
      .remove(() => true)
      .insert(data))
    .run();

  console.log(
    'ITER',  ++count,
    'DATA',  view._runtime.data.livedata.values.value.length,
    'ITEMS', view.scenegraph().root.items[0].items[2].items.length,
    'CACHE', view._runtime.nodes[79].value.size,
    'TIME',  Date.now() - t0
  );
  t0 = Date.now();
}, 100);


// remove history future back then 'period' for time ordered data
function timeWindow(data, period) {
    var now = Date.now();
    var i = 0;
    while (now - data[i].time > period && i < data.length) {
        i++;
    }
    return data.slice(i, data.length);
}

// data producer
setInterval(function() {
    data = timeWindow(data, 31e3);
    data.push({y: data[data.length-1].y + Math.random() * 10 - 5, time: Date.now()});
}, 10);

update();
    </script>
  </body>
</html>
```

**Comment by jheer** [comment]
Thanks! This helped me pinpoint the issue: your change sets sometimes include the _same_ tuple in both the add and remove bins. Vega assumes these sets are distinct, so mixing them can lead to faulty tracking. To get a working example, you can instead queue up new tuples and add those only. Doing so actually simplifies your logic, too!

Here is a relevant snippet, modifying your example:

```js
var data = [{y:0, time: Date.now()}],
    last = data[0], // most recent data point
    period = 31e3;

// update data every 100ms
setInterval(function() {
  var t1 = Date.now();

  view
    .change('livedata', view.changeset()
      .remove(d => (t1 - d.time) > period)
      .insert(data))
    .run();

  data = []; // clear new data buffer
}, 100);

// data producer
setInterval(function() {
  last = {y: last.y + Math.random() * 10 - 5, time: Date.now()};
  data.push(last);
}, 10);
```

**Comment by jheer** [comment]
FYI, I've added a new issue (#1318) for adding safeguards to changeset processing to catch issues like these proactively.

**Comment by rmcgregor1990** [comment]
My actual use case for these live graphs is within a React-Redux application. All the datasets are live in the Redux store, so not being able to cleanly override vegas dataset on a re-render is a bit sad. 

I guess can perform some from of diff of the new and the old datasets to derive the changeset.

Cheers for taking to the time to look into this,
Rob

**Comment by jheer** [comment]
I agree. I just published [vega-dataflow 4.0.2](https://github.com/vega/vega-dataflow/releases/tag/v4.0.2), which includes new safeguards for changeset processing. Your example code now works with these new changes, which will be included in the Vega 4.0.0 release.

**Comment by rmcgregor1990** [comment]
Awesome, thanks!
